packages/core/src/utils/security/index.ts

```diff
@@ -0,0 +1,54 @@
+import { isLeft, isRight, Left } from 'fp-ts/lib/Either';
+import { pipe } from 'fp-ts/lib/pipeable';
+import { get, identity } from 'lodash';
+import { set } from 'lodash/fp';
+import { securitySchemaHandlers } from './handlers';
+import { AuthErr, SecurityScheme } from './handlers/types';
+
+import * as Either from 'fp-ts/lib/Either';
+
+function getAllInvalidSec<R>(invalidSecuritySchemas: Array<Left<AuthErr>>) {
+  const pathToHeader = ['headers', 'WWW-Authenticate'];
+
+  const firstLeftValue: AuthErr = pipe(
+    invalidSecuritySchemas[0],
+    Either.fold<AuthErr, R, AuthErr>(identity, identity),
+  );
+
+  return firstLeftValue.status !== 401 || invalidSecuritySchemas.length === 1
+    ? firstLeftValue
+    : (() => {
+        const allWWWAuthHeaders = invalidSecuritySchemas.reduce((accumulator: string, currentValue) => {
+          return pipe(
+            currentValue,
+            Either.mapLeft(err => [accumulator, get(err, pathToHeader)].filter(identity).join(', ')),
+            Either.fold(authHeader => authHeader || '', () => ''),
+          );
+        }, '');
+
+        return set(pathToHeader, allWWWAuthHeaders, firstLeftValue);
```

Comment:
Hmm, can you quickly explain what’s going on here?

Reply:
Sure :) so, first, all `WWW-Authenticate` headers were gathered from schemes that did not pass the validation: `allWWWAuthHeaders` (https://github.com/stoplightio/prism/blob/f68e4abe51d9c7284dc30ee28974acfcbdeaa810/packages/core/src/utils/security/index.ts#L21).

Then, I've taken the first scheme that failed - `firstLeftValue` (the value of the first Left) and "extended" its `['headers', 'WWW-Authenticate']` to be `allWWWAuthHeaders`. So the response will have that header with info on ways to authenticate.